src/Components/Server/src/Circuits/CircuitHost.cs

```diff
@@ -589,6 +589,30 @@ public void SendPendingBatches()
         _ = Renderer.Dispatcher.InvokeAsync(Renderer.ProcessBufferedRenderBatches);
     }
 
+    private Task DispatchInboundActionAsync(Func<Task> func)
+    {
+        return DispatchInboundActionAsyncCore(0, func);
+    }
+
+    private async Task<TResult> DispatchInboundActionAsync<TResult>(Func<Task<TResult>> func)
+    {
+        TResult result = default;
+        await DispatchInboundActionAsyncCore(0, async () => result = await func());
+        return result;
+    }
+
+    private async Task DispatchInboundActionAsyncCore(int circuitHandlerIndex, Func<Task> func)
+    {
+        if (circuitHandlerIndex >= _circuitHandlers.Length)
+        {
+            await func();
+            return;
+        }
+
+        var circuitHandler = _circuitHandlers[circuitHandlerIndex];
+        await circuitHandler.HandleInboundActivityAsync(Circuit, () => DispatchInboundActionAsyncCore(circuitHandlerIndex + 1, func));
+    }
+
```

Comment:
Do you think we can avoid the recursion here?

Also, do you think we can build the handler chain "once" instead of on a per event basis?